cmd/disk-cache.go

```diff
@@ -0,0 +1,674 @@
+func (c diskCache) encodedPath(bucket, object string) string {
+	return path.Join(c.dataDir, fmt.Sprintf("%x", sha256.Sum256([]byte(path.Join(bucket, object)))))
+}
+
+// encode (bucket,object) to a hash value which is used as the file name of the cached object's metadata.
+func (c diskCache) encodedMetaPath(bucket, object string) string {
+	return path.Join(c.dataDir, fmt.Sprintf("%x.json", sha256.Sum256([]byte(path.Join(bucket, object)))))
+}
+
+// Commit the cached object - rename from tmp directory to data directory.
+func (c diskCache) Commit(f *os.File, objInfo ObjectInfo, anon bool) error {
+	bucket := objInfo.Bucket
+	object := objInfo.Name
+
+	objectLock := globalNSMutex.NewNSLock(diskCacheLockingPrefix+bucket, object)
+	objectLock.Lock()
+	defer objectLock.Unlock()
+
+	encPath := c.encodedPath(bucket, object)
+	if err := os.Rename(f.Name(), encPath); err != nil {
+		return err
+	}
+	if err := f.Close(); err != nil {
+		return err
+	}
+	metaPath := c.encodedMetaPath(bucket, object)
+	objMeta := newDiskCacheObjectMeta(objInfo, anon)
+	metaBytes, err := json.Marshal(objMeta)
+	if err != nil {
+		return err
+	}
+
+	if err = ioutil.WriteFile(metaPath, metaBytes, 0644); err != nil {
+		return err
+	}
+	c.updateAtime(bucket, object, time.Now().UTC())
+	return nil
+}
+
+// Remove the cached object from the tmp directory.
+func (c diskCache) NoCommit(f *os.File) error {
+	tmpName := f.Name()
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	return os.Remove(tmpName)
+}
+
+// Creates a file in the tmp directory to which the cached object is written to.
+// Once the object is written to disk the caller can call diskCache.Commit()
+func (c diskCache) Put(size int64) (*os.File, error) {
+	if c.diskUsageHigh() {
+		select {
+		case c.purgeChan <- struct{}{}:
+		default:
+		}
+		return nil, errDiskFull
+	}
+	f, err := ioutil.TempFile(c.tmpDir, "")
+	if err != nil {
+		return nil, err
+	}
+	e := Fallocate(int(f.Fd()), 0, size)
+	switch {
+	case isSysErrNoSys(e), isSysErrOpNotSupported(e), e == nil:
+		// Ignore errors when Fallocate is not supported in the current system
+		err = nil
+	case isSysErrNoSpace(e):
+		err = errDiskFull
+	default:
+		err = e
+	}
+	if err != nil {
+		tmpPath := f.Name()
+		f.Close()
+		os.Remove(tmpPath)
+		return nil, err
+	}
+	return f, nil
+}
+
+// Returns the handle for the cached object
+func (c diskCache) Get(bucket, object string) (*os.File, ObjectInfo, bool, error) {
+	objectLock := globalNSMutex.NewNSLock(diskCacheLockingPrefix+bucket, object)
+	objectLock.RLock()
+	defer objectLock.RUnlock()
+
+	metaPath := c.encodedMetaPath(bucket, object)
+	objMeta := diskCacheObjectMeta{}
+	metaBytes, err := ioutil.ReadFile(metaPath)
+	if err != nil {
+		return nil, ObjectInfo{}, false, err
+	}
+	if err = json.Unmarshal(metaBytes, &objMeta); err != nil {
+		return nil, ObjectInfo{}, false, err
+	}
+	if objMeta.Version != diskCacheObjectMetaVersion {
+		return nil, ObjectInfo{}, false, errors.New("format not supported")
```

Comment:
Use `errCorruptedFormat`

Reply:
errCorruptedFormat will not be right because it gives the impression that the format file is itself corrupted. But this is a case where older binary running on newer formats.

Reply:
Then use `errFormatNotSupported`